### PR TITLE
Fix #6169 - DataTable: Lazy getRowData Repeat the query twice

### LINKED ISSUE #6169: DataTable: Lazy getRowData Repeat the query twice
**1) Environment**

PrimeFaces version: 8.0

org.joinfaces.primefaces-spring-boot-starter  4.3.0

**2) Expected behavior**

I hope that when I select a piece of data in the datatable, getRowData is executed only once.

**3) Actual behavior**

When I select a piece of data in the datatable, getRowData is executed twice, which is very strange.

**AbstractLazyModel**

```
    public abstract class AbstractLazyModel<T> {

    private T[] selects;
    private T select;

    private LazyDataModel<T> lazyDataModel;

    protected abstract LambdaQueryWrapper<T> getQueryParam();

    protected abstract BasicService getBasicService();

    protected  abstract String getKey(T t);


    public LazyDataModel<T> getResultModel() {
        if (lazyDataModel == null) {
            lazyDataModel = new LazyDataModel<T>() {
                @Override
                public List<T> load(int first, int pageSize, Map<String, SortMeta> sortMeta, Map<String, FilterMeta> filterMeta) {
                    int currentPage = (first / pageSize) + 1;
                    Page<T> page =  new Page(currentPage,pageSize);
                    page.setOptimizeCountSql(false);
                    return getBasicService().getResultList(page,getQueryParam());
                }

                @Override
                public T getRowData(String rowKey) {
                    return   (T)getBasicService().findById(rowKey);
                }

                @Override
                public String getRowKey(T t) {
                    return getKey(t);
                }
            };
            lazyDataModel.setRowCount( getBasicService().getRowCount(getQueryParam()));
        }
        return lazyDataModel;
    }

    protected void refresh() {
        this.setSelect(null);
        search();
    }

    public void search() {
        lazyDataModel = null;
    }
```

**Example Bean**

```
@Slf4j
@Data
@Component
@ViewScoped
public class ErpSysCodeTypeView extends AbstractLazyModel<ErpSysCodeType> implements java.io.Serializable {

    @Autowired
    private ErpSysCodeTypeService erpSysCodeTypeService;


    @Autowired
    private PrimeFaces primeFaces;

    @Autowired
    private FacesContext facesContext;

    private SysCodeTypeRequest sysCodeTypeRequest;

    private String codeName;


    private List<ErpSysCodeType> sysCodeTypeList;

    private ErpSysCodeType oldErpSysCodeType;


    public void searchListener() {
        this.refresh();
    }

    public void onRowSelect(SelectEvent<ErpSysCodeType> event) {
      // this.setSelect( event.getObject());
    }




    @PreDestroy
    public void destroy() {
        log.info("ErpSysCodeTypeView destroy");
    }


    @Override
    protected LambdaQueryWrapper<ErpSysCodeType> getQueryParam() {
        return new LambdaQueryWrapper<ErpSysCodeType>().
                likeRight(!StringUtils.isEmpty(codeName),ErpSysCodeType::getName,codeName);
    }


    @Override
    protected BasicService getBasicService() {
        return erpSysCodeTypeService;
    }

    @Override
    protected String getKey(ErpSysCodeType erpSysCodeType) {
        return String.valueOf(erpSysCodeType.getId());
    }

}

```

**Example Xhtml**

```
  <p:dataTable id="dataTableId"
                                         widgetVar="sysCodeTypeTableVar"
                                         value="#{erpSysCodeTypeView.resultModel}"
                                         paginatorPosition="bottom"
                                         rowIndexVar="rowIdx"
                                         rows="10"
                                         lazy="true"
                                         selection="#{erpSysCodeTypeView.select}"
                                         selectionMode="single"
                                         rowKey="#{_sysCodeType.id}"
                                         var="_sysCodeType"
                                         emptyMessage="暂无数据."
                                         paginator="true"
                                         currentPageReportTemplate="查询到{totalRecords}条记录，本页显示{startRecord}～{endRecord}；第{currentPage}/{totalPages}页"
                                         paginatorTemplate="{CurrentPageReport} {FirstPageLink} {PreviousPageLink} {PageLinks} {NextPageLink} {LastPageLink} {RowsPerPageDropdown}"
                                         rowsPerPageTemplate="5,10,15">

                                <p:ajax event="rowSelect"  listener="#{erpSysCodeTypeView.onRowSelect}" process="@this"/>

                                <p:column width="50">
                                    <f:facet name="header">序号</f:facet>
                                    <h:outputText value="#{rowIdx + 1}"/>
                                </p:column>

                                <p:column headerText="字典代码">
                                    <h:outputText value="#{_sysCodeType.code}"/>
                                </p:column>

                                <p:column headerText="字典名称">
                                    <h:outputText value="#{_sysCodeType.name}"/>
                                </p:column>

                                <p:column headerText="启用">
                                    <h:outputText value="#{_sysCodeType.status == 1 ? '启用':'关闭'}"/>
                                </p:column>

                                <p:column headerText="创建时间">
                                    <h:outputText value="#{_sysCodeType.createDate}">
                                        <f:convertDateTime pattern="yyyy-MM-dd HH::mm:ss"/>
                                    </h:outputText>
                                </p:column>

                            </p:dataTable>

```

### PR DESCRIPTION
**Pull request by Rapster**
New PR replacing #6571

>@melloware Can you explain your solution? Is this just moving all the code around so the method doesn't get called from the datatable and the feature?

> @Rapster  Fair point! I'm going to start with this: there is no way to prevent calling `DataTable#getRowData`. My first attempt was to use a cache to prevent this and have better perf, but as the model is reseted all the time (by design I'm guessing) my PR didn't bring a lot to the table.
> 
> Now, my idea is to avoid scenarios like expensive call to differents endpoints (e.g database, REST etc.), turns out the result sounds like a lot like SelectItemsConverter from Omnifaces does, except it's not based on toString() representation but either `DataTable#rowKey` or `System#identityHashCode`.
> 
> To sum up:
> - users no longer need to implement `LazyDataModel#getRowKey` and `LazyDataModel#getRowData`. As of today, either you need to set `DataTable#rowKey` or `LazyDataModel#getRowKey` if lazy mode, from now on one entry point: `DataTable#rowKey` and yet, this is optional.
> 
> - Whenever selection has to be decoded, `LazyDataModel#getRowData` will under the hood, loop over Datatable rows till it finds a matching result by using rowKey strategy that has been set with `DataTable#rowKey`. You might wonder what is that cache in DefaultSelectableModel, let's say i select all elements in reverse order (from bottom to top) time complexity will remain O(N), today this scenario is O(N²) _(because it'll keep iterating over elements even though some have been visited)_
> 
> Back to the issue, no query will be executed at all

### CODE REVIEW

**Comment by melloware** [comment]
`RowKeyAware` seems to be a name that is consistent with our other interfaces like `TouchAware` `FlexAware` etc?

**Comment by Rapster** [comment]
I thought of that too, but most of these `xxxAware` are used for components while it's not the case here so I didn't want create any confusion. Being said, we could have a `RowKeyAware` extending `RowKeyAble` which adds `getRowKey()` method. But so far, we only have datatable like this, Tree/TreeTable handle selection differently, so I don't think it worth it for now...

**Comment by melloware** [comment]
No I agree its not worth introducing another Interface was just trying to come up with a better name than `RowKeyAble`

**Comment by Rapster** [comment]
What about  `RowKeyMapper`?

**Comment by melloware** [comment]
I like `RowKeyMapper` better than `Able`!

**Comment by tandraschko** [comment]
JFYI: I likely wont review + apply it for 9.0, we are already in bugfix mode here and holiday time is near.

**Comment by Rapster** [comment]
I think it would be nice to have this for next release, as it's a good addition for the component, and users will be happy to no longer implements getRowKey and getRowData. If @tandraschko you don't have time, maybe someone could have a try in case I missed something?

**Comment by tandraschko** [comment]
Im unsure if we should support identityHashCode as it doesnt work most of the times.
It only works when your model is "longer scoped" then requestscoped and no LazyDataModel is used. IMO its better to force a rowKey.

**Comment by Rapster** [comment]
As long your `DataTable#value` doesn't change every request, _(and I don't know about how other users do it)_, but in my case whatever holds data is ViewScoped, so it will work all the times. But even if it's RequestScoped, it's just a default rowkey to ease your devs, nothing forces you to use that key, you can always set your own rowKey in such scenario (and in some cases like MVS, it's actually the thing to do, see doc)

**Comment by tandraschko** [comment]
it just will only work for one single case:
`List<Model>` which is in View or SessionScoped

everything else should be broken and `List<Model>` in a longer scope is probably the worst implementation a user could do.

**Comment by Rapster** [comment]
You mean `List<Model>` in a ViewScoped bean is the worst implementation a user could do?

**Comment by tandraschko** [comment]
in most of the times, yes
the source of a DataTable is mostly from a Database and therefore LazyDataModel is the way to go.
Otherwise you have "all pages" and rows saved in the view and/or session, which bad for memory and often replicated to other nodes.

**Comment by Rapster** [comment]
I could use `Object#hashCode` as default actually, it's more deterministic _(if Object#hashCode implemented of course)_ Obviously, it might not work for cell editing, but just like MVS, `DataTable#rowKey` needs to be defined and as long users are aware of that it's good enough

**Comment by tandraschko** [comment]
IMO we should align the documentation that rowKey can be implemented to use hashCode in the applicaitno-code for this basic usecase and we should still force rowKey to be implemented.

hashCode is of course a better here but, as you already said, it only works if implemented. And when we still force rowKey to be implemented in the application, the users are aware that hashCode/or custom rowKey needs to be implemented correctly.

**Comment by tandraschko** [comment]
@Rapster i will revert this for now. We didnt finish the discussion yet, so please just dont merge the PR next time.

**Comment by tandraschko** [comment]
We could also think about making rowkeymapper a converter

**Comment by Rapster** [comment]
Ok, what's wrong?

**Comment by tandraschko** [comment]
i only reviewed the hashCode part of this whole PR and we didnt even agree on this single part yet ;D

**Comment by tandraschko** [comment]
some general thoughts:

1) it would be great if you could write more text on a PR, its hard to follow a bigger refactoring just by reading code
2) RowKeyMapper can be replaced by Converter, AFAIR we had a issue about it
3) LazyDataModel#getRowData really belongs to LazyDataModel as it it lazy-loads a single Entity
4) RowKeyMapperDataModel, even not sure why you introduced it, has a cache which can hold wrong data therefore we had 3)

**Comment by Rapster** [comment]
> it would be great if you could write more text on a PR, its hard to follow a bigger refactoring just by reading code

Sorry I didn't realize it was that big... If I'm not mistaken, I explained the whole solution in the original PR

> RowKeyMapper can be replaced by Converter, AFAIR we had a issue about it

I think you're refering to [this](https://github.com/primefaces/primefaces/issues/223). I thought about it, indeed it has both methods similar to getRowKey and getRowData but since `SelectableDataModel` was already there, I followed the same idea. I can tweak something using `Converter`

> LazyDataModel#getRowData really belongs to `LazyDataModel` as it it lazy-loads a single Entity

Not really, `LazyDataModel` is all about fetching data depending certain criterias and nothing more. Making the correspondence between an `Object` and a `String` representation _(used for selection for example)_ is the responsability of something else like a Converter/RowKeyMapper. That's why I came up with `RowKeyMapperDataModel`, there is a clean separation between who's responsible for handling selection, and another for loading data. But also, there is two entry point for setting rowKey, `DataTable#rowKey` and `LazyDataModel#getRowKey` which make it more confusing to use the feature. `DataTable#getRowKey` should be the only entry point. In other words, whether it's lazy or not, it goes through the same code: `RowKeyMapperDataModel`

> RowKeyMapperDataModel, even not sure why you introduced it, has a cache which can hold wrong data therefore we had 3)

Explained in the orignal PR, in two words, it's for performance reasons _(I'm going to rename `cache` to avoid any confusion)_. One thing, I could improve on that front is as to when I should clear it up, I reverted that commit... But if you have a closer look, you'll realize `UIData#setDataModel(null)` is constantly called, so that's why I didn't really bothered as to when I should clear it up. So holding wrong data is very unlikely to happen, specially that it's only used for decoding selection (String to Object)

P.S: I attached the explanation of the solution from the original PR in here (which I should've probably done in the first place 😅 )

**Comment by melloware** [comment]
@Rapster yes that is the original issue if you do support converter let's reopen that ticket and link the PR.

**Comment by Rapster** [comment]
@melloware Basically what I have in mind so far, is just to make `RowKeyMapperDataModel` implements `Converter`, and nothing more _(RowKeyMapper becomes useless, therefore removed)_. I don't intend to support converters for `DataTable` for this use case. @djmj You might have an opinion also on this PR.

**Comment by melloware** [comment]
Yep that is fine with me!
